### PR TITLE
Data validation import

### PR DESCRIPTION
**Pull request by reiterl**

### CODE REVIEW

**Comment by reiterl** [comment]
This is raw and work in progress. It is a large change, so I give other developers a change to look at it.

File: openslides_backend/models/check_models.py [hunk 1]
```diff
@@ -0,0 +1,686 @@
...
+            if collection not in collection_allowlist:
+                del self.models[collection]
+        # TODO: mediafile blob handling.
+
+    def generate_template_prefixes(self) -> None:
+        for collection in self.models.keys():
+            for field in self.get_fields(collection):
+                if not isinstance(field, BaseTemplateField):
+                    continue
+                field_name = field.get_structured_field_name("")
```
Comment (FinnStutzenstein) [inline openslides_backend/models/check_models.py:184]: You should be able to use `BaseTemplateField.index` instead of splitting the structured field by yourself.

File: openslides_backend/models/check_models.py [hunk 2]
```diff
@@ -0,0 +1,686 @@
...
+        if self.is_structured_field(field):
+            field, _ = self.to_template_field(collection, field)
+
+        field_type = cast(BaseRelationField, self.models[collection]().get_field(field))
+        return (
+            field_type.get_target_collection().collection,
+            field_type.to.get(field_type.get_target_collection()),
+        )
+
+    def find_model(self, collection: str, id: int) -> Optional[Dict[str, Any]]:
```
Comment (FinnStutzenstein) [inline openslides_backend/models/check_models.py:549]: Maybe create a new issue: This might be not performant for many models of one collection -> One can create a dict mapping `id` to the model for quick access.

File: tests/system/action/meeting/test_import.py [hunk 3]
```diff
@@ -23,6 +23,158 @@ def create_request_data(self, datapart: Dict[str, Any]) -> Dict[str, Any]:
                         "projector_countdown_warning_time": 60,
                         "reference_projector_id": 1,
                         "user_ids": [1],
+                        "imported_at": None,
```
Comment (FinnStutzenstein) [inline tests/system/action/meeting/test_import.py:26]: Can you use https://github.com/OpenSlides/OpenSlides/blob/openslides4-dev/docs/example-data.json [1], load it once and only do your modifications for each test?

[1] You can copy this file next to the test and remove unwanted collections (organization, committee, ...)

**Review by FinnStutzenstein — CHANGES_REQUESTED** [review]

File: openslides_backend/models/check_models.py (hunk 1 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py:184]: I use the Field function here, it should be okay.

File: tests/system/action/meeting/test_import.py (hunk 3 above)
Comment (reiterl) [reply tests/system/action/meeting/test_import.py:26]: No, that would be not good. The tests work by the way.

File: openslides_backend/models/check_models.py (hunk 2 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py:549]: Add an issue: https://github.com/OpenSlides/openslides-backend/issues/827

**Comment by FinnStutzenstein** [comment]
Please create a PR in the main repo which deletes the data validator and adapts the CI

File: openslides_backend/models/check_models.py [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+
+color_regex = re.compile("^#[0-9a-f]{6}$")
+
+
+def check_color(value: Any) -> bool:
+    return value is None or bool(isinstance(value, str) and color_regex.match(value))
+
+
+def check_number(value: Any) -> bool:
+    return value is None or isinstance(value, int)
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:70]: This is not up-to-date: https://github.com/OpenSlides/OpenSlides/blob/77f6516438c93e790fd2915a32d416bb3a0f88a5/docs/datavalidator/check_json.py#L53

File: openslides_backend/models/check_models.py [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+
+
+def check_float(value: Any) -> bool:
+    return value is None or isinstance(value, int) or isinstance(value, float)
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:74]: same as above

File: openslides_backend/models/check_models.py [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+        Only checks that for each replacement a structured field exists and
+        not too many structured fields. Does not check the content.
+        Returns True on errors.
+        """
+        errors = False
+        for template_field in self.get_fields(collection):
+            if not isinstance(template_field, BaseTemplateField):
+                continue
+            field_error = False
+            replacements = model.get(template_field.get_structured_field_name(""))
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:314]:
```suggestion
            replacements = model.get(template_field.get_template_field_name())
```

File: openslides_backend/models/check_models.py (hunk 1 above)
Comment (jsangmeister) [reply openslides_backend/models/check_models.py:185]:
```suggestion
                field_name = field.get_template_field_name()
```

File: openslides_backend/models/check_models.py [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+        collection_fields_types = (
+            x
+            for x in self.models[collection]().get_fields()
+            if self.is_normal_field(x.own_field_name)
+            or self.is_template_field(x.own_field_name)
+        )
+        collection_fields = set(
+            field.own_field_name
+            if not isinstance(field, BaseTemplateField)
+            else field.get_structured_field_name("")
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:288]:
```suggestion
            else field.get_template_field_name()
```

File: openslides_backend/models/check_models.py [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+        errors = False
+        for template_field in self.get_fields(collection):
+            if not isinstance(template_field, BaseTemplateField):
+                continue
+            field_error = False
+            replacements = model.get(template_field.get_structured_field_name(""))
+
+            if not isinstance(replacements, list):
+                self.errors.append(
+                    f"{collection}/{model['id']}/{template_field.own_field_name}: Replacements for the template field must be a list"
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:318]: `own_field_name` contains the name without the `$`, better use `get_template_field_name()`

File: openslides_backend/models/check_models.py [hunk 9]
```diff
@@ -0,0 +1,690 @@
...
+                )
+                field_error = True
+                continue
+            for replacement in replacements:
+                if not isinstance(replacement, str):
+                    self.errors.append(
+                        f"{collection}/{model['id']}/{template_field.own_field_name}: Each replacement for the template field must be a string"
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:325]: `own_field_name` contains the name without the `$`, better use `get_template_field_name()`

File: openslides_backend/models/check_models.py [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+            if template_field.replacement_collection:
+                replacement_collection = (
+                    template_field.replacement_collection.collection
+                )
+
+            for replacement in replacements:
+                structured_field = self.make_structured(template_field, replacement)
+                if structured_field not in model:
+                    self.errors.append(
+                        f"{collection}/{model['id']}/{template_field.own_field_name}: Missing {structured_field} since it is given as a replacement"
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:341]: `own_field_name` contains the name without the `$`, better use `get_template_field_name()`

File: openslides_backend/models/check_models.py [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+                    )
+                    errors = True
+
+                if replacement_collection:
+                    try:
+                        as_id = int(replacement)
+                    except (TypeError, ValueError):
+                        self.errors.append(
+                            f"{collection}/{model['id']}/{template_field.own_field_name}: Replacement {replacement} is not an integer"
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:350]: `own_field_name` contains the name without the `$`, better use `get_template_field_name()`

File: openslides_backend/models/check_models.py [hunk 12]
```diff
@@ -0,0 +1,690 @@
...
+                        )
+                    if not self.find_model(replacement_collection, as_id):
+                        self.errors.append(
+                            f"{collection}/{model['id']}/{template_field.own_field_name}: Replacement {replacement} does not exist as a model of collection {replacement_collection}"
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:354]: `own_field_name` contains the name without the `$`, better use `get_template_field_name()`

File: openslides_backend/models/check_models.py [hunk 13]
```diff
@@ -0,0 +1,690 @@
...
+
+            for field in model.keys():
+                if self.is_structured_field(field):
+                    try:
+                        _template_field, _replacement = self.to_template_field(
+                            collection, field
+                        )
+                        if (
+                            template_field.own_field_name == _template_field
+                            and _replacement not in model[template_field.own_field_name]
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:365]: Here it is actually relevant: `model[template_field.own_field_name]` will never exist because of the missing `$`

File: openslides_backend/models/check_models.py [hunk 14]
```diff
@@ -0,0 +1,690 @@
...
+                        ):
+                            self.errors.append(
+                                f"{collection}/{model['id']}/{field}: Invalid structured field. Missing replacement {_replacement} in {template_field.own_field_name}"
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:368]: see above

File: openslides_backend/models/check_models.py [hunk 15]
```diff
@@ -0,0 +1,690 @@
...
+    def check_types(self, model: Dict[str, Any], collection: str) -> None:
+        for field in model.keys():
+            if self.is_template_field(field):
+                continue
+
+            field_type = self.get_type_from_collection(field, collection)
+            enum = self.get_enum_from_collection_field(field, collection)
+
+            checker: Optional[Callable[..., bool]] = None
+            if isinstance(field_type, CharField):
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:388]: Why did you remove the mapping? I think it is cleaner to list it there and only access it here (the mapping would be of type `Dict[Field, Callable]` now, of course)

File: openslides_backend/models/check_models.py [hunk 16]
```diff
@@ -0,0 +1,690 @@
...
+
+    def split_collectionfield(self, collectionfield: str) -> Tuple[str, str]:
+        collection, field = collectionfield.split("/")
+        if collection not in self.models.keys():
+            raise CheckException(
+                f"Collectionfield {collectionfield} has an invalid collection"
+            )
+        if field not in [
+            field.own_field_name for field in self.models[collection]().get_fields()
+        ]:  # Note: this has to be adopted when supporting template fields
```
Comment (jsangmeister) [inline openslides_backend/models/check_models.py:640]: Please resolve this TODO: for template fields, `get_template_field_name` must be used (maybe you should add a util method like `get_own_field_name` or similar, since this pattern seems to occur more often)

File: openslides_backend/action/actions/meeting/import_.py [hunk 17]
```diff
@@ -159,15 +138,12 @@ def update_meeting_users_and_mediafiles(self, instance: Dict[str, Any]) -> None:
...
 
-        # delete blob from mediafiles
-        for entry in json_data.get("mediafile", []):
-            if "blob" in entry:
-                del entry["blob"]
-
     def create_replace_map(self, json_data: Dict[str, Any]) -> None:
         replace_map: Dict[str, Dict[int, int]] = {}
         for collection in json_data:
             replace_map[collection] = {}
```
Comment (jsangmeister) [inline openslides_backend/action/actions/meeting/import_.py:144]: you could use a defaultdict here and remove this line

**Review by jsangmeister — CHANGES_REQUESTED** [review]

File: openslides_backend/models/check_models.py (hunk 1 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py:185]: Replaced.

File: openslides_backend/models/check_models.py (hunk 6 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py:314]: Replaced.

File: openslides_backend/models/check_models.py (hunk 7 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py:288]: Replaced.

File: openslides_backend/models/check_models.py (hunk 15 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py:388]: I got problems with the type checking and tried Dict[Field, Callable]. It did not work.

File: openslides_backend/models/check_models.py (hunk 4 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py:70]: Updated.

File: openslides_backend/models/check_models.py (hunk 5 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py:74]: Updated.

File: openslides_backend/models/check_models.py (hunk 16 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py:640]: Add a function `get_own_field_name` and use it instead `field.own_field_name`

File: openslides_backend/models/check_models.py (hunk 15 above)
Comment (jsangmeister) [reply openslides_backend/models/check_models.py:387]: Doesn't mypy tell you what type it actually is? If you can't figure it out, you can still put `Any` or `# type: ignore` there, but restructuring the code because you can't figure out the correct type is usually not the correct approach.

File: openslides_backend/models/check_models.py (hunk 16 above)
Comment (jsangmeister) [reply openslides_backend/models/check_models.py:640]: Please remove the TODO then

File: openslides_backend/action/actions/meeting/import_.py (hunk 17 above)
Comment (reiterl) [reply openslides_backend/action/actions/meeting/import_.py:144]: Which variable should be a defaultdict?

Comment (jsangmeister) [reply openslides_backend/action/actions/meeting/import_.py:144]: `replace_map`

Comment (reiterl) [reply openslides_backend/action/actions/meeting/import_.py:144]: Ah, okay, you mean the `replace_map`.

Comment (reiterl) [reply openslides_backend/action/actions/meeting/import_.py]: Done.

File: openslides_backend/models/check_models.py (hunk 8 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py]: Use get_own_field_name here.

File: openslides_backend/models/check_models.py (hunk 9 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py]: Use get_own_field_name here.

File: openslides_backend/models/check_models.py (hunk 10 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py]: Use get_own_field_name here.

File: openslides_backend/models/check_models.py (hunk 11 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py]: Use get_own_field_name here.

File: openslides_backend/models/check_models.py (hunk 12 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py]: Use get_own_field_name here.

File: openslides_backend/models/check_models.py (hunk 13 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py]: Use get_own_field_name here.

File: openslides_backend/models/check_models.py (hunk 14 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py]: Use get_own_field_name here.

File: openslides_backend/models/check_models.py (hunk 15 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py:387]: Use now Any, now it is okay. Use the checker_map again.

File: openslides_backend/models/check_models.py (hunk 16 above)
Comment (reiterl) [reply openslides_backend/models/check_models.py:640]: Removed

**Comment by jsangmeister** [comment]
I added better typing and fixed an error for the checker map. Also I extracted the callable part of the checker into its own file in `cli/` since that seemed more appropriate. Let me know what you think. If that's fine with you @reiterl you can go ahead and merge this.

**Comment by reiterl** [comment]
Seems okay for me.
I also thought about splitting the file into cli and the check_type typing fix seems to work without the Template Fields. Okay.

**Comment by FinnStutzenstein** [comment]
This is ok for me. It would be nice, if the CI in the main repo will be changed together with this PR; @jsangmeister can you do this?

**Comment by jsangmeister** [comment]
> This is ok for me. It would be nice, if the CI in the main repo will be changed together with this PR; @jsangmeister can you do this?

Yes
